pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,449 @@
+		}
+
+		// Calculate the round-trip time
+		end := time.Now()
+		rtt := end.Sub(sentTime)
+
+		// Update the latency store
+		mutator := func(entry *NodeIPLatencyEntry) {
+			entry.LastSendTime = sentTime
+			entry.LastRecvTime = end
+			entry.LastMeasuredRTT = rtt
+		}
+		m.latencyStore.SetNodeIPLatencyEntry(destIP, mutator)
+	}
+}
+
+// pingAll sends ICMP messages to all the Nodes.
+func (m *NodeLatencyMonitor) pingAll(ipv4Socket, ipv6Socket net.PacketConn) {
+	klog.V(4).InfoS("Pinging all Nodes")
+	nodeIPs := m.latencyStore.ListNodeIPs()
+	for _, toIP := range nodeIPs {
+		if toIP.To4() != nil && ipv4Socket != nil {
+			if err := m.sendPing(ipv4Socket, toIP); err != nil {
+				klog.ErrorS(nil, "Cannot send ICMP message to Node IP because socket is not initialized for IPv4", "IP", toIP)
+			}
+		} else if toIP.To16() != nil && ipv6Socket != nil {
+			if err := m.sendPing(ipv6Socket, toIP); err != nil {
+				klog.ErrorS(nil, "Cannot send ICMP message to Node IP because socket is not initialized for IPv6", "IP", toIP)
+			}
+		} else {
+			klog.ErrorS(nil, "Cannot send ICMP message to Node IP because socket is not initialized for IP family", "IP", toIP)
+		}
+	}
+	klog.V(4).InfoS("Done pinging all Nodes")
+}
+
+// Run starts the NodeLatencyMonitor.
+func (m *NodeLatencyMonitor) Run(stopCh <-chan struct{}) {
+	go m.nodeLatencyMonitorInformer.Informer().Run(stopCh)
+	go m.nodeInformer.Informer().Run(stopCh)
+	go m.monitorLoop(stopCh)
+
+	<-stopCh
+}
+
+// monitorLoop is the main loop to monitor the latency of the Node.
+func (m *NodeLatencyMonitor) monitorLoop(stopCh <-chan struct{}) {
+	klog.V(4).InfoS("NodeLatencyMonitor is running")
+	// Low level goroutine to handle ping loop
+	var ticker *time.Ticker
+	var tickerCh <-chan time.Time
+	var ipv4Socket, ipv6Socket net.PacketConn
+	var err error
+
+	defer func() {
+		if ipv4Socket != nil {
+			ipv4Socket.Close()
+		}
+		if ipv6Socket != nil {
+			ipv6Socket.Close()
+		}
+		if ticker != nil {
+			ticker.Stop()
+		}
+	}()
+
+	// Update current ticker based on the latencyConfig
+	updateTicker := func(interval time.Duration) {
+		if ticker != nil {
+			ticker.Stop() // Stop the current ticker
+		}
+		ticker = time.NewTicker(interval)
+		tickerCh = ticker.C
+	}
+
+	klog.V(4).InfoS("NodeLatencyMonitor is running2")
+	wg := sync.WaitGroup{}
+	// Start the pingAll goroutine
+	for {
+		select {
+		case <-tickerCh:
+			// Try to send pingAll signal
+			m.pingAll(ipv4Socket, ipv6Socket)
+		case <-stopCh:
+			return
+		case <-m.latencyConfigChanged:
+			// Start or stop the pingAll goroutine based on the latencyConfig
+			if m.latencyConfig.Enable {
+				// If the Nodes is empty(after deleting the CRD), we need to wait for the NodeInformer to populate the Nodes.
+				if len(m.latencyStore.ListNodeIPs()) == 0 {
+					nodes, err := m.nodeInformer.Lister().List(labels.Everything())
+					if err != nil {
+						klog.ErrorS(err, "Failed to list Nodes")
+					}
+					m.latencyStore.addNodes(nodes)
+				}
```

Comment:
Actually this looks like the wrong approach.
When the CRD is deleted, the informers are still running and we are still processing Node ADD / UPDATE / DELETE events. This means that the latencyStore will be updated even when the monitor is stopped.
I think that we should remove this code (along with the new `addNodes` method), and remove the `m.latencyStore.CleanUp()` call later in this function.

(We can keep the rest of the deletion logic as is, including the `WaitGroup`).

Reply:
Ok, @Dyanngg have mentioned in the previous comment that we need to clear all the map contents when deleting the CRD to save memory, so with this List and CleanUp method, I'll update it.

Reply:
We should keep the maps as they are and see if there is room for optimization later.
Right now the code is clunky, and we can end up with a strange behavior when Nodes are added / deleted after the latency monitor is stopped.
For example, if a Node is deleted during that time, this will logged an error:
```
	nodeIPs, ok := l.nodeTargetIPsMap[node.Name]
	if !ok {
		klog.ErrorS(nil, "Failed to get IPs for Node", "nodeName", node.Name)
		return
	}
```
We are talking about a very small amount of memory here, because we only store the latest latency measurement. So it's not worth making the code more complex and potentially incorrect.

Reply:
Okay, I'll remove those implementations.